Prevent deadlocks from touch callbacks on chained associations

avit opened the pull request:
Even with great care to arrange our model updates consistently within
transaction blocks, ActiveRecord could still cause deadlocks based on the
sequence of models needing to touch associations before commit.

The order that records are appended to the transaction callback array should be
sorted in a sequence that is:

1. Consistent between different transactions operating on the same tables, and
2. Least likely to conflict with the order of statements across different tables

Consider a simple hierarchy of Child belongs to Parent with touch: true. We can
easily create a situation where the deferred touches are listed out of order on
transaction records, based on different data conditions:

Transaction A

    Child.nice.each(&:send_gifts)
    # => [
    #   #<Child id: 1, parent_id: 1>,
    #   #<Child id: 2, parent_id: 2>
    # ]
    current_transaction.records
    # => [
    #   #<Parent id: 1>
    #   #<Parent id: 2>
    # ]

Transaction B

    Child.naughty.each(&:send_coal)
    # => [
    #   #<Child id: 3, parent_id: 2>,
    #   #<Child id: 4, parent_id: 1>
    # ]
    current_transaction.records
    # => [
    #   #<Parent id: 2>
    #   #<Parent id: 1>
    # ]

If these two transactions were to run concurrently, the `before_commit` callback
sequence could end up as:

|     | Transaction A             | Transaction B             |               |
| --- | ---                       | ---                       | ---           |
| 1   | UPDATE parents WHERE id=1 |                           | A locks row 1 |
| 2   |                           | UPDATE parents WHERE id=2 | B locks row 2 |
| 3   | UPDATE parents WHERE id=2 |                           | A waits for B |
| 4   |                           | UPDATE parents WHERE id=1 | B waits for A |

As each transaction is holding a row lock and waiting on the other to complete,
the database has no choice but to ROLLBACK one of these transactions with a
corresponding deadlock error.

To prevent this, we sort the transaction callback records in order of:

1. *Greatest depth* so when different associations reference a parent through
different paths, the deepest dependencies are applied first. This mostly mimics
the current implementation as associations get traversed upwards. (Polymorphic
associations are not counted for depth since they are indeterminate.)
2. *Table name* so different models that refer to the same table are applied
together in sequence, and not interleaved with other models.
3. *Primary key* as records from the same table can be listed for touch
callbacks in different order, they need to be updated in sequence.

avit committed:
Prevent deadlocks from touch callbacks on chained associations

Even with great care to arrange our model updates consistently within
transaction blocks, ActiveRecord could still cause deadlocks based on the
sequence of models needing to touch associations before commit.

The order that records are appended to the transaction callback array should be
sorted in a sequence that is:

1. Consistent between different transactions operating on the same tables, and
2. Least likely to conflict with the order of statements across different tables

Consider a simple hierarchy of Child belongs to Parent with touch: true. We can
easily create a situation where the deferred touches are listed out of order on
transaction records, based on different data conditions:

Transaction A

    Child.nice.each(&:send_gifts)
    # => [
    #   #<Child id: 1, parent_id: 1>,
    #   #<Child id: 2, parent_id: 2>
    # ]
    current_transaction.records
    # => [
    #   #<Parent id: 1>
    #   #<Parent id: 2>
    # ]

Transaction B

    Child.naughty.each(&:send_coal)
    # => [
    #   #<Child id: 3, parent_id: 2>,
    #   #<Child id: 4, parent_id: 1>
    # ]
    current_transaction.records
    # => [
    #   #<Parent id: 2>
    #   #<Parent id: 1>
    # ]

If these two transactions were to run concurrently, the `before_commit` callback
sequence could end up as:

|     | Transaction A             | Transaction B             |               |
| --- | ---                       | ---                       | ---           |
| 1   | UPDATE parents WHERE id=1 |                           | A locks row 1 |
| 2   |                           | UPDATE parents WHERE id=2 | B locks row 2 |
| 3   | UPDATE parents WHERE id=2 |                           | A waits for B |
| 4   |                           | UPDATE parents WHERE id=1 | B waits for A |

As each transaction is holding a row lock and waiting on the other to complete,
the database has no choice but to ROLLBACK one of these transactions with a
corresponding deadlock error.

To prevent this, we sort the transaction callback records in order of:

1. *Greatest depth* so when different associations reference a parent through
different paths, the deepest dependencies are applied first. This mostly mimics
the current implementation as associations get traversed upwards. (Polymorphic
associations are not counted for depth since they are indeterminate.)
2. *Table name* so different models that refer to the same table are applied
together in sequence, and not interleaved with other models.
3. *Primary key* as records from the same table can be listed for touch
callbacks in different order, they need to be updated in sequence.

avit commented:
Updated from previous implementation #33900

rafaelfranca commented on activerecord/test/cases/touch_later_test.rb:
```diff
@@ -120,4 +121,89 @@ def test_touching_three_deep
+    end
+
+    setup do
+      Tree.prepend WaitForConcurrentCommit
```
Would not this leak to other tests? I think it is better to create subclasses of those that are only used on this test.

avit replied:
Yes but it's a no-op unless the instance attribute is set. I could add it to the actual test model, or else subclass it as you suggest.

rafaelfranca reviewed (commented):
This is now doing way more work than the previous implementation. How often those deadlocks happen with regular associations> Do this problem only happen with `touch: true` associations?

avit commented:
I was still drafting this, but let's mark it ready for review since we're into the discussion now. :)

> This is now doing way more work than the previous implementation

It's still a straightforward sort, and the depth lookup is cached on the class. I had some [questions](https://github.com/rails/rails/pull/33900#issuecomment-458020028) on the original PR about this concern originally. The key points for doing the extra work is:

- It's only called once per transaction (i.e. ± per request or job)
- Touch options only apply to singular associations, so the array size can only be on the order of O(n) for records updated in the block
- It's going to reduce load from server error retries

I found the extra depth sorting to be necessary because sorting only by the class or table names as before could still have potential for deadlocks.

Effectively, update statements are sent to the DB in two phases: first during the transaction block, and then the `before_commit` callback array for `touch_later`. The depth sorting is needed to avoid overlap between the two phases.

The current rails implementation _does_ follow depth as the order that associations are traversed when they are added to the callback array—but it ignores the case of multiple rows from the same table, which is what was addressed in the first PR. (Same-table deadlocks seemed to be much more common for us so it was still a huge improvement.)

To extract a real-world example from our application, we have a belongs-to hierarchy that looks like `Requirement` → `Feature` → `Epic` with touches for nested caching. Now consider two transactions, one that touches the bottom of the hierarchy (a requirement) and the other only touches the middle (a feature) for the same 3 records.

This is what Rails currently does (correctly):

```
# BEGIN                                   # BEGIN

req.update!(some_attrs)                   feature.update!(some_attrs)
# UPDATE requirements                     # UPDATE features

transaction.records << req.feature        transaction.records << feature.epic
transaction.records << req.feature.epic  

# -- touch before commit                  # -- touch before commit

# UPDATE features                         # UPDATE epics
# UPDATE epics

# COMMIT                                  # COMMIT
```

If we only sort the transaction callback records by model or table name, then we get:

```
# BEGIN                                   # BEGIN
# UPDATE requirements                     # UPDATE features

# -- touch before commit (sorted by name):

# UPDATE epics                            # UPDATE epics
# UPDATE features
```

On the left, *E* before *F* can deadlock with *Feature* then *Epic*. (The need for this is harder to illustrate in the actual Rails test suite—unless we want to create new table names like `z_nodes` and `a_trees`.)

However, we still need to sort records within the same tables consistently, as illustrated by the *children* example in the PR description. So yes, we should do both.

> How often those deadlocks happen with regular associations?

This very much depends on request frequency for the application, and with more background workers (resque/sidekiq/activejob) that likelihood increases. We were seeing it a lot with the rails gem implementation before last year. We've been running the implementation from #33900 for over a year now, and that has reduced the problem dramatically (I'd say over 95%). I'm really hoping this new implementation eliminates the rest.

The remainder is still frequent enough that we are addressing ourselves with this patch.

> Do this problem only happen with touch: true associations?

Yes, TouchLater is the only thing I see in the default ActiveRecord `before_commit` callback stack.

This would also change the order that records are called back for any other user-defined `before_commit` hooks, so if authors are emitting any update statements in those, then I think this consistent sorting could only help there, too.

avit committed:
Remove lazy enum as it's actually slower

avit committed:
Simplify logic for final comparison

To handle potential of new records with nil primary key. This is cleaner than
`&.nonzero?`

avit committed:
Fix depth traversal: each branch up the tree needs its own stack

This was memoizing incorrect depth values when there were multiple association
paths pointing to a parent & grandparent association. If the shorter path was
traversed first, it could be ignored by the deeper association.

eugeneius commented on activerecord/test/cases/touch_later_test.rb:
```diff
@@ -120,4 +121,101 @@ def test_touching_three_deep
+      self.table_name = "nodes"
+
+      belongs_to :tree, class_name: ConcurrentTree.name, touch: true
+      belongs_to :parent, class_name: self.name, touch: true
```
The deadlock in `test_touch_callback_deadlocks_by_association_depth` is caused by the order in which these associations are declared: updating a second-level node touches the tree first (moving up two levels in the association hierarchy) and then its parent node (moving down a level again). I don't think that Rails should reorder these touches automatically, given that the developer can fix the issue by changing the order of the associations.

More generally: as long as the application ensures that touches only cascade in one direction, we can avoid the problem described in https://github.com/rails/rails/pull/38146#issuecomment-570694381 without reflecting on the model's associations by preserving the order that tables first appear in the `records` array:

```patch
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb b/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
index 7893ec85d6..aea1bd8371 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb
@@ -116,7 +116,13 @@ def rollback_records
       end

       def before_commit_records
-        records.uniq.each(&:before_committed!) if records && @run_commit_callbacks
+        return unless records && @run_commit_callbacks
+
+        records.uniq.group_by do |record|
+          record.class.table_name
+        end.each_value do |records|
+          records.sort_by(&:id).each(&:before_committed!)
+        end
       end

       def commit_records
```

avit replied:
Yes, the declaration order within one model is obviously one way of causing this. But assuming the author will need to take that into account seems less "declarative" and much harder to trace and prove correct between different models.

Something very general like `belongs_to :user` (depth=1) is most typically going to be the first thing declared on several common models above other associations (that themselves may indirectly belong to other users).

> as long as the application ensures that touches only cascade in one direction

That's a very reasonable assumption and right now I only perform the depth lookups for `belongs_to` as well. But thinking about this, there's no reason we couldn't also consider `has_one` depths for completeness. (Still, I would assume those are edge-cases in any model tree. Typically the deepest upward traversal would be via `belongs_to`.)

I did consider using `group_by` to sort by the first occurence of each table instead of name order. However, that doesn't promise the order will be the same between different transaction blocks.

It very much depends on the data and runtime conditions, too. Just for one example, associations can be optional:

Transaction A:

```
Child.create(parent_id: nil, school_id: 1)
Child.create(parent_id: 1, school_id: nil)

transaction.records == [#<School id: 1>, #<Parent id: 1>]
```

Transaction B:

```
Child.create(parent_id: 1, school_id: nil)
Child.create(parent_id: nil, school_id: 1)

transaction.records == [#<Parent id: 1>, #<School id: 1>]
```

I would still recommend that ActiveRecord should apply some constant, deterministic ordering to these callbacks.

avit replied:
@eugeneius I could rewrite the test with the associations declared in your suggested "correct" order, and provide additional scenarios for the other examples I showed here.

I don't think the `group_by` sorting is sufficient so maybe adding a test case to fail that implementation would be helpful.

Please let me know how I can help move this forward.

avit replied:
@eugeneius is there still any interest in this patch?

Let me know so I can update this.

avit commented:
I've just published a blog post that explains the need for this patch:

[Transaction Deadlocks on ActiveRecord associations][1]

[1]: https://www.aha.io/engineering/articles/transaction-deadlocks-in-activerecord

taylorthurlow commented:
I'm noticing that this issue has gone stale, which I think is a bit of a shame, seeing as no real conclusion has been reached about whether or not this PR is something the core team wants to merge. I don't have a horse in this race but I'd rather see this issue closed by someone with a wontfix label than the bot closing it as stale, especially when someone puts this amount of effort into explaining the problem and explaining the solution in-depth.

schneidmaster commented:
Any chance someone from the Rails team can provide feedback on this? I don't mean be annoying as I know from personal experience that open source maintenance is challenging and thankless, but we really feel that this would be a significant improvement for a lot of Rails users. In our application this took us from thousands of deadlocks a month down to almost zero -- deadlocks were a tiny portion of our overall traffic but that's still thousands of experiences where a user's work inexplicably failed to save or other buggy behavior that is totally avoidable. Once your application scales to ~hundreds of models it's very difficult to mentally map what order each belongs_to needs to be in to avoid deadlocks, and it wasn't clear in our case that such an ordering was even possible. We did not notice any measurable performance impact from this change, but even if there was, it seems completely in line with the Rails philosophy to accept slightly slower performance in exchange for an intuitive developer experience, and avoiding developers banging their head against the wall trying to figure out deadlocks that inevitably crop up once you reach a certain scale. Again, not trying to be a bother, but Andrew put an enormous amount of work into researching and implementing this solution (as well as the accompanying blog post) so we would love to hear if you have any interest or feedback.

neohunter commented:
I agree, this should be seriously considered to be included into ActiveRecord. Deadlock issues are very hard to trace and asking developers  to declare relationships keeping the same order is annoying (or better, expecting that they guess without any information).

neohunter commented:
@avit one question, it's safe to do monkey patching with the files you changed?

Something like adding this in a initializer https://gist.github.com/neohunter/91dbe6031d099527bbc6c5a75e51d505

avit commented:
@neohunter Yes, that is effectively how we are patching this in production right now.

avit commented:
@eileencodes would this fix make sense to include for Rails 6.1 or is the door closed on that already? I'm only asking since I saw the 6.1-rc1 announcement today and it's been 10 months since I submitted this PR.

neohunter commented:
From our experience, including this fix greatly improved the performance and reduce random errors that our customers was having, we are losing customers and money each week. We grew to a point where this errors was happening to majority of our customers without being able to fix this, we had a team of 6 developers working on differents performance improvements and this change reduced the deadlocks in a 90% and "random" errors that our customers was having. 

This greatly improves the experience of everybody. 

- Our sales teams can schedule demos without worrying 
- Our customers can use with confidence our system again
- Our developers can work in "real" features, instead of trying to fix a very strange and har to follow up issue
- The customers of our customers stop complaining with them.

taylorthurlow commented:
!unstale

dvandersluis commented:
@avit is there a way to try this out since it's been over a year with no response? We get a ton of deadlocks in production and I would love to see if this fixes our issue.

taylorthurlow commented:
!unstale

Alexander-Senko commented:
@avit, could you please release the fix as a gem? They gonna be years until it's merged otherwise.

taylorthurlow commented:
!unstale

tenderlove commented:
Is sorting by depth necessary?  If sorting just by name and id isn't sufficient, could you explain why and add a test demonstrating it?  Thanks for sending this PR, and sorry it's taking so long to be resolved.

avit commented:
Hi Aaron, thanks for weighing in. I outlined the reason for the depth sorting in my [earlier comment](https://github.com/rails/rails/pull/38146#issuecomment-570694381) above.

I'm trying to think of what tables in the existing test suite would make sense to use for a test here, or if I should add a new set of tables for "requirements", "features", "epics" specifically for this. (The trick here is that "epics" comes before "features" if sorted alphabetically, the reverse of what a depth sort would do.)

avit commented:
Meanwhile, working on writing some illustrative test cases...

---

> In other words, this isn't actually possible to fix generally speaking. Do I understand that correctly?

If "this" generally means any regular `save` operation with associations, I believe the fix is actually possible with what we're doing here:

- Make ActiveRecord::TouchLater put association records in a deterministic order
- Despite loading different data, every `before_commit` will update its records in the same order
- Transactions will wait for the database if they touch the same record
- Because of consistent order, we avoid A waiting on B while B is waiting on A

Yes, it's still possible to write app code that accidentally causes deadlocks between different transactions by ordering operations inconsistently _in app code_, but now it's something that can be fixed, unlike the framework callback phase which was previously non-deterministic.

---

To bring everything up to date for 2022, in our own real-world usage at Aha! we found:

1. The previous implementation (simple sort) actually reduced deadlocks by over 95%.
2. We then identified some complex transactions where updates were being applied in inconsistent order and fixed those ourselves in app code, sometimes needing to use an earlier `lock!` to force things around.
3. We found other places where the simple sort approach was too simple, having introduced new deadlocks because it clobbered Rails' original depth traversal. Hence this improved sort.
4. We now have literally **zeros** of deadlocks that can be attributed to ActiveRecord touches since then.

eileencodes commented:
Hey @avit, thanks for the PR. @tenderlove and I are looking at reducing some of the complexity with a refactor, but we're having trouble with some of the conditionals. There are a bunch of branches in these conditionals that are not tested so it's difficult to ensure that the refactoring is going in the right direction.

It appears there are no tests for the following cases:

* `belonging.polymorphic?` (https://github.com/rails/rails/pull/38146/files#diff-dd0ea9dfea1fd486f2b1b5eac8c798f839467173a1c2315d3ad04f4098f7dda3R352)
* `stack.include?(table_name)` (https://github.com/rails/rails/pull/38146/files#diff-dd0ea9dfea1fd486f2b1b5eac8c798f839467173a1c2315d3ad04f4098f7dda3R354) - I split up the conditional and saw a puts only for tests in the table equality check `belonging.table_name == model.table_name`
* When the else is greater than 1 (tests also pass if I hardcode 0 for the association depth value) https://github.com/rails/rails/blob/09ebaf966b5a1c237ae43d7b612e228be38dfb2e/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L357

Would you be able to help write tests for these cases so we can be sure everything is covered?

avit commented:
Thanks for the review @eileencodes. You're right there's nothing in the current test suite to cover those cases. I might have to set up some new tables for these scenarios, which I am aware of, and then maybe we can see if it makes sense to fold any of them back into the existing ActiveRecord test schema.

I'll work on this.

---

`belonging.polymorphic?` is a level of belonging so we know it's not `0` but we don't clearly know to what model unless we make the lookup dynamic so we take a shortcut and just say `1`. The depth is determined by the longest belonging chain on the model, so for testing this we will need a model with:

1. Only a polymorphic belongs-to, else its depth would be zero
2. A table name lower in the alphabet than the belonging parent table[^1]

---

`stack.include?(table_name)` covers circular references. Testing scenario could be:

1. one model that belongs to a tree node
2. a different (grand)parent model that is also linked to the same tree structure

---

The `else` association depth there is recursive so we look up the belonging depth and add 1 for the current model. For this scenario I think we need a grandparent structure to compare a depth 2 model against a depth 1.

1. A grandparent model named "epics" (depth 0)
2. A parent model named "features" (depth 1)
3. A child model named "requirements" or "tasks" (depth 2)

(These are our real-world table names that exhibit the sorting conflict. We could just as easily name them `a_`, `b_` and `c_` but it might not be as clear.)

[^1]: lower means towards "z", to prove that alphabet sorting of the table names won't cover this.

guilhermemaranhao commented:
Hi, @avit, @eileencodes, how is the current state of the `touch` option on polymorphic tables, concerning the deadlock issue? I'm on `rails 7.1.3.2`

rafaelfranca commented:
The code changed significantly since this was open. Sorry about that. @avit do you want to update the PR with current code? I'll merge after that
